### PR TITLE
Added RobotState Object

### PR DESCRIPTION
**Pull request by jared-paul**
### Description
Added a RobotState abstraction to remove a lot of circular buffers from robot.cpp. This shouldn't change any underlying implementation, and has no new implementation to test.

### Testing Done
Built and ran all tests, everything passed. Since this doesn't add any new logic, refactoring the old tests to use this state class instead of the circular buffers should be sufficient.

### Resolved Issues
#807 

### Length Justification
A lot of classes use the robot class as it was so I had to refactor all of the existing implementation.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by jared-paul** [comment]
Forgot to create a new branch for this so it contains all my old commits from BallState, sorry

File: src/software/world/robot.cpp
Comment (scveloso) [inline src/software/world/robot.cpp]: Should this be ```Trying to update **robot** state using a state older than the current state``` ?

File: src/software/ai/evaluation/deflect_off_enemy_target.cpp
Comment (scveloso) [inline src/software/ai/evaluation/deflect_off_enemy_target.cpp]: Instead of constructing a new ```RobotState``` with the Robot attributes, might be better to use ```i.currentState()``` ?

File: src/software/world/team.cpp
Comment (scveloso) [inline src/software/world/team.cpp]: Same thing, maybe can use ```robot.currentState()``` here?

**Review by scveloso — CHANGES_REQUESTED** [review]
Looking good, just some minor comments

File: src/software/ai/evaluation/deflect_off_enemy_target.cpp
Comment (garethellis0) [reply src/software/ai/evaluation/deflect_off_enemy_target.cpp]: Hmm, seems like we would want the enemy history here as well, so why not `enemyClosestToEdge = i;`?

Comment (garethellis0) [inline src/software/ai/evaluation/deflect_off_enemy_target.cpp]: nit, optional; both `enemyClosestToEdge` and `i` violate our naming conventions, `enemyClosestToEdge` should be `enemy_closest_to_edge` and `i` should be `enemy_robot` (or really anything more descriptive then `i`). Out-of-scope, so feel free to ignore, but would be good to fix them.

File: src/software/world/ball_state.cpp
Comment (garethellis0) [inline src/software/world/ball_state.cpp:32]: :eyes: :heavy_check_mark:

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looking good, no major comments!

File: src/software/world/robot_state.h
Comment (jonathanlew) [inline src/software/world/robot_state.h]: Can we add a class comment here?

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Nothing much to add beyond what's already been said :+1:

File: src/software/world/robot_state.h
Comment (jared-paul) [reply src/software/world/robot_state.h]: done

File: src/software/world/ball_state.cpp
Comment (jared-paul) [reply src/software/world/ball_state.cpp:32]: ya... :eyes: my bad from the BallState issue

File: src/software/ai/evaluation/deflect_off_enemy_target.cpp
Comment (jared-paul) [reply src/software/ai/evaluation/deflect_off_enemy_target.cpp]: done

Comment (jared-paul) [reply src/software/ai/evaluation/deflect_off_enemy_target.cpp]: done

File: src/software/world/robot.cpp
Comment (jared-paul) [reply src/software/world/robot.cpp]: done

**Comment by jared-paul** [comment]
note: CI will fail due to #886

File: src/software/world/robot_state.h
Comment (jonathanlew) [reply src/software/world/robot_state.h]: @jared-paul Can we add a comment describing what RobotState is? For example, RobotState represents the position, velocity, orientation, and angular velocity of a robot at a particular point in time.

File: src/software/world/ball_state.cpp
Comment (garethellis0) [reply src/software/world/ball_state.cpp:32]: All good! We've all put in fixes like this :laughing:

File: src/software/world/robot_state.h
Comment (garethellis0) [reply src/software/world/robot_state.h]: Perhaps a little more generic then that, just because at that point you're basically just repeating the class members?

**Comment by garethellis0** [comment]
I'm confused, why does #886 break this? Will we be good once merge conflicts are resolved?

**Comment by MathewMacDougall** [comment]
> I'm confused, why does #886 break this? Will we be good once merge conflicts are resolved?

It's an uninitialized value so _can_ break it, and happens to in CI right now just because of how memory is layed out.

File: src/software/ai/evaluation/deflect_off_enemy_target.cpp
Comment (MathewMacDougall) [inline src/software/ai/evaluation/deflect_off_enemy_target.cpp]: Should be `shortest_len_to_edge` according to our style guide

File: src/software/ai/evaluation/deflect_off_enemy_target.cpp
Comment (MathewMacDougall) [inline src/software/ai/evaluation/deflect_off_enemy_target.cpp]: Should be `dir_perp` according to our style guide

File: src/software/backend/output/grsim/command_primitive_visitor/motion_controller_test.cpp
Comment (MathewMacDougall) [inline src/software/backend/output/grsim/command_primitive_visitor/motion_controller_test.cpp]: Should be `temp_robo` according to our style guide

File: src/software/backend/output/grsim/command_primitive_visitor/motion_controller_test.cpp
Comment (MathewMacDougall) [inline src/software/backend/output/grsim/command_primitive_visitor/motion_controller_test.cpp]: Should be `temp_robo` according to our style guide

File: src/software/world/robot.cpp
Comment (MathewMacDougall) [inline src/software/world/robot.cpp]: So interesting point here. Because we are providing an `unsigned int` it's actually impossible for this value to be negative. However it can be an incorrect positive value depending how the compiler wraps values.

I think just having `unsigned int` would be enough, since once compiler errors are enabled implicit conversions can't happen to we are protected that way. Currently this code really serves no purpose and can't even be tested. Thoughts?

**Edit**
This should probably check if the size is `< 1`. What happens if there is no history? All the getter functions would segfault.

File: src/software/world/robot.cpp
Comment (MathewMacDougall) [inline src/software/world/robot.cpp]: I think we don't really need this function? It kinda defeats the point of having the `State` class since it just exposes all the same members.

File: src/software/world/robot.cpp
Comment (MathewMacDougall) [inline src/software/world/robot.cpp]: Do we want to keep these `getPrevious<stuff>` functions still now that we have state? I think I would prefer only having a `getPreviousStates` function instead of individual ones for each state member variable?

File: src/software/world/robot.h
Comment (MathewMacDougall) [inline src/software/world/robot.h]: semi-nit: What are everyone's thoughts on `updateState` vs `updateCurrentState`? Seeing it changed to `updateCurrentState` threw me off for a second because I thought there might also be a `updateState` function and wasn't sure what the difference would be. IMO `updateCurrentState` almost implied the current state is mutated directly? This might all be due to my previous expectations and just being used to the old names, so would really like some input here.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
sorry for some of the nitpicking, but overall this is looking great :+1:

File: src/software/world/robot_state.h
Comment (MathewMacDougall) [reply src/software/world/robot_state.h]: Perhaps

"This class represents the state of a Robot at a single point in time"?

Comment (garethellis0) [reply src/software/world/robot_state.h]: :+1: I like that.

**Comment by MathewMacDougall** [comment]
@jared-paul any updates on this?

File: src/software/world/robot.cpp
Comment (MathewMacDougall) [reply src/software/world/robot.cpp]: @garethellis0 @jonathanlew @scveloso thoughts on this?

File: src/software/world/robot.cpp
Comment (MathewMacDougall) [reply src/software/world/robot.cpp]: @garethellis0 @scveloso @jonathanlew thoughts on this?

File: src/software/world/robot.h
Comment (MathewMacDougall) [reply src/software/world/robot.h]: @garethellis0 @jonathanlew @scveloso thoughts on this?

**Comment by MathewMacDougall** [comment]
@garethellis0 @jonathanlew @scveloso please re-review.

Also @jared-paul looks like formatting needs to be fixed.

File: src/software/world/robot.cpp
Comment (garethellis0) [reply src/software/world/robot.cpp]: Agreed.

File: src/software/world/robot.cpp
Comment (garethellis0) [reply src/software/world/robot.cpp]: I'm fine with this being public.

File: src/software/world/robot.h
Comment (garethellis0) [reply src/software/world/robot.h]: I like `updateState`, not a fan of `updateCurrentState`.

Comment (scveloso) [reply src/software/world/robot.h]: Yeah I feel like ```updateState``` kind of already implies update the current state

File: src/software/world/robot.cpp
Comment (scveloso) [reply src/software/world/robot.cpp]: +1 on only having ```getPreviousStates```

File: src/software/world/robot.cpp
Comment (scveloso) [reply src/software/world/robot.cpp]: Agree, I think caller should be using a RobotState if caller has those fields anyway

**Review by garethellis0 — COMMENTED** [review]
LGTM pending other comments here

**Comment by MathewMacDougall** [comment]
@jared-paul any updates on this?

**Comment by jared-paul** [comment]
@MathewMacDougall I can't seem to get formatting to pass CI, when I run the script it doesn't change any files, will have this done by tomorrow

**Comment by garethellis0** [comment]
@jared-paul - are you running `./clang-formatting/fix-formatting -b master`? If so, can you update your local version of master to the latest on upstream? For the sake of speed, using `-b master` will diff your branch with master and only format the bits that you're changing. @MathewMacDougall - we should consider having the script diff against `upstream/master` instead of local `master` instead?

**Comment by MathewMacDougall** [comment]
> we should consider having the script diff against upstream/master instead of local master instead?

Yes absolutely, and making that the default?

**Comment by jared-paul** [comment]
@garethellis0 @MathewMacDougall I've tried "../clang_format/fix_formatting -a", "../clang_format/fix_formatting -b master", pulling from upstream master then running "../clang_format/fix_formatting -a"

I don't know why this is happening

**Comment by garethellis0** [comment]
So I pulled from upstream master and then ran `./clang_format/fix_formatting.sh -b master` on this branch and got:
```
changed
files:
src/software/backend/output/grsim/command_primitive_visitor/motion_controller_test.cpp
 
========================================================================
clang-format has modified the above files. Formatting complete.
```
Could you try the same and confirm you still don't get any files changed?

**Comment by jared-paul** [comment]
@garethellis0 I did "git pull upstream master" in this branch ran "./clang_format/fix_formatting.sh -b master" and it said no files needed to be formatted

**Comment by garethellis0** [comment]
uhh, let's take a look in-person tomorrow, not sure what's up

File: src/software/world/robot.cpp
Comment (jonathanlew) [reply src/software/world/robot.cpp]: I'm OK with it being public

File: src/software/world/robot.cpp
Comment (jonathanlew) [reply src/software/world/robot.cpp]: Agreed

File: src/software/world/robot.h
Comment (jonathanlew) [reply src/software/world/robot.h]: Agreed

Comment (jared-paul) [reply src/software/world/robot.h]: changed updateCurrentState to updateState

File: src/software/world/robot.cpp
Comment (jared-paul) [reply src/software/world/robot.cpp]: refactored to ball and robot to have only getPreviousStates

File: src/software/world/robot.cpp
Comment (jared-paul) [reply src/software/world/robot.cpp]: removed this one in favour of just having updateState(RobotState state)

**Review by garethellis0 — DISMISSED** [review]
LGTM - pending other comments

**Review by jonathanlew — DISMISSED** [review]

**Review by MathewMacDougall — DISMISSED** [review]
LGTM :+1:

**Comment by MathewMacDougall** [comment]
@scveloso please re-review

**Comment by MathewMacDougall** [comment]
@jared-paul sorry I know you just had an exam, but do you have time to wrap this up? I started working on the `PhysicsRobot` class for the simulator and I'm going to have to deal with merge conflicts, so would rather do that sooner than later.

**Comment by jared-paul** [comment]
@MathewMacDougall Ya I'll wrap it up now, (hopefully it builds successfully). I completely thought this was already merged haha, my bad

**Comment by MathewMacDougall** [comment]
Looks like there's a build error, otherwise LGTM

**Comment by MathewMacDougall** [comment]
@jared-paul ^

**Comment by jared-paul** [comment]
@MathewMacDougall Sorry, but the tests "test_assigning_2_robots_to_2_tactics_with_overlap", "test_assigning_3_robots_to_3_tactics_with_2_of_the_same_cost", and "test_assigning_2_robots_to_1_tactic_unsatisfied_robotcapabilityflags" keep failing for me locally

**Comment by jared-paul** [comment]
@MathewMacDougall nevermind, I just ran a series of git commands that I think shouldn't have affected it at all and now it passes lol

**Review by MathewMacDougall — DISMISSED** [review]
:eyes: :+1: :tada: :fire: :100: :relieved:

**Comment by MathewMacDougall** [comment]
@scveloso please approve if your changes have been addressed.

**Comment by MathewMacDougall** [comment]
Looks like there's some conflicts now @jared-paul :grimacing:

**Comment by jared-paul** [comment]
@MathewMacDougall Yea :( , I can't merge with that requested change though

**Comment by MathewMacDougall** [comment]
@jared-paul Why? Can you resolve the merge manually?

**Comment by jared-paul** [comment]
@MathewMacDougall ya but github won't let me merge until all the requested changes are resolved I think

**Comment by MathewMacDougall** [comment]
Oh yea you can't merge the PR because it's blocked on @scveloso 's approval still. I was referring to the merge conflicts that have appeared due to other PRs getting merged.

**Comment by garethellis0** [comment]
@jared-paul - just to be clear here, we need to:
- resolve merge conflicts
- @scveloso needs to give approval

These two things are independent, you can resolve merge conflicts without @scveloso's approval?

**Comment by jared-paul** [comment]
@garethellis0 Yep! I will resolve these today

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada:

**Comment by MathewMacDougall** [comment]
@scveloso please re-review

**Review by MathewMacDougall — DISMISSED** [review]

**Review by MathewMacDougall — APPROVED** [review]

**Comment by MathewMacDougall** [comment]
@scveloso Please re-review

**Review by scveloso — APPROVED** [review]

**Pull request merged**
